pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,594 @@
+	// Deprecated, but required for back-compat (including e2e)
+	deprecatedPublicIPs []string
+}
+
+// returns a new serviceInfo struct
+func newServiceInfo(service proxy.ServicePortName) *serviceInfo {
+	return &serviceInfo{
+		sessionAffinityType: api.ServiceAffinityNone, // default
+		stickyMaxAgeMinutes: 180,                     // TODO: paramaterize this in the API.
+	}
+}
+
+// Proxier is an iptables based proxy for connections between a localhost:lport
+// and services that provide the actual backends.
+type Proxier struct {
+	mu         sync.Mutex // protects serviceMap
+	serviceMap map[proxy.ServicePortName]*serviceInfo
+	iptables   utiliptables.Interface
+}
+
+// Proxier implements ProxyProvider
+var _ proxy.ProxyProvider = &Proxier{}
+
+// NewProxier returns a new Proxier given an iptables Interface instance.
+// Because of the iptables logic, it is assumed that there is only a single Proxier active on a machine.
+// An error will be returned if iptables fails to update or acquire the initial lock.
+// Once a proxier is created, it will keep iptables up to date in the background and
+// will not terminate if a particular iptables call fails.
+func NewProxier(ipt utiliptables.Interface) (*Proxier, error) {
+	glog.V(2).Info("Tearing down userspace rules.")
+	// remove iptables rules/chains from the userspace Proxier
+	tearDownUserspaceIptables(ipt)
+	return &Proxier{
+		serviceMap: make(map[proxy.ServicePortName]*serviceInfo),
+		iptables:   ipt,
+	}, nil
+}
+
+// Chains from the userspace proxy
+// TODO: Remove these Chains and tearDownUserspaceIptables once the userspace Proxier has been removed.
+var iptablesContainerPortalChain utiliptables.Chain = "KUBE-PORTALS-CONTAINER"
+var iptablesHostPortalChain utiliptables.Chain = "KUBE-PORTALS-HOST"
+var iptablesContainerNodePortChain utiliptables.Chain = "KUBE-NODEPORT-CONTAINER"
+var iptablesHostNodePortChain utiliptables.Chain = "KUBE-NODEPORT-HOST"
+
+// tearDownUserspaceIptables removes all iptables rules and chains created by the userspace Proxier
+func tearDownUserspaceIptables(ipt utiliptables.Interface) {
+	// NOTE: Warning, this needs to be kept in sync with the userspace Proxier,
+	// we want to ensure we remove all of the iptables rules it creates.
+	// Currently they are all in iptablesInit()
+	// Delete Rules first, then Flush and Delete Chains
+	args := []string{"-m", "comment", "--comment", "handle ClusterIPs; NOTE: this must be before the NodePort rules"}
+	if err := ipt.DeleteRule(utiliptables.TableNAT, utiliptables.ChainOutput, append(args, "-j", string(iptablesHostPortalChain))...); err != nil {
+		glog.Errorf("Error removing userspace rule: %v", err)
+	}
+	if err := ipt.DeleteRule(utiliptables.TableNAT, utiliptables.ChainPrerouting, append(args, "-j", string(iptablesContainerPortalChain))...); err != nil {
+		glog.Errorf("Error removing userspace rule: %v", err)
+	}
+	if err := ipt.FlushChain(utiliptables.TableNAT, iptablesContainerPortalChain); err != nil {
```

Comment:
this would read better as
```
chains []iptables.Chain{iptablesContainerPortalChain, iptablesHostPortalChain, iptablesHostNodePortChain, iptablesContainerNodePortChain}
for _, c := range chains {
    ipt.FlushChain(utiliptables.TableNAT, c)
    //error handling 
   ipt.DeleteChain(utiliptables.TableNAT, c)
   // error handling
}